iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java

```diff
@@ -493,18 +503,42 @@ private String getStatsSource() {
   }
 
   private Path getStatsPath(Table table) {
-    return new Path(table.location() + STATS + table.name() + table.currentSnapshot().snapshotId());
+    return getStatsPath(table, table.currentSnapshot().snapshotId());
   }
 
-  private void invalidateStats(Path statsPath) {
+  private Path getStatsPath(Table table, long snapshotId) {
+    return new Path(table.location() + STATS + table.name() + snapshotId);
+  }
+
+  private boolean checkAndInvalidateStats(Table tbl) {
+    Path statsPath = getStatsPath(tbl);
     try {
       FileSystem fs = statsPath.getFileSystem(conf);
       if (fs.exists(statsPath)) {
-        fs.delete(statsPath, true);
+        // Analyze table and stats updater thread
+        return fs.delete(statsPath, true);
       }
     } catch (IOException e) {
       LOG.error("Failed to invalidate stale column stats: {}", e.getMessage());
     }
+    return false;
+  }
+
+  private void checkAndMergeStats(ColumnStatistics statsObjNew, Table tbl,
+      org.apache.hadoop.hive.ql.metadata.Table hmsTable, ColumnStatsDesc columnStatsDesc) {
+    Long previousSnapshotId = tbl.currentSnapshot().parentId();
+    if (previousSnapshotId != null && canProvideColStatistics(tbl, previousSnapshotId)) {
+      ColumnStatistics statsObjOld = readColStats(tbl, getStatsPath(tbl, previousSnapshotId));
+      if (statsObjOld != null && statsObjOld.getStatsObjSize() != 0 && !statsObjNew.getStatsObj().isEmpty()) {
+        try {
+          MetaStoreServerUtils.mergeColStats(statsObjNew, statsObjOld);
+          StatsSetupConst.setColumnStatsState(hmsTable.getParameters(), columnStatsDesc.getColName());
+        } catch (InvalidObjectException e) {
+          StatsSetupConst.removeColumnStatsState(hmsTable.getParameters(), columnStatsDesc.getColName());
```

Comment:
Why do we need to set or remove the column stats state here? I think setting it in a single place depending upon whether statistics is correctly written as done in #4440 should suffice right?

Reply:
If there is an exception,  merging stats won't complete as a result the stats would be inaccurate.

Reply:
wait, we persist stats only after the successful merge, then why do we need `removeColumnStatsState`?

Reply:
fixed, we will persist stats only after the successful merge and return true. If unsuccessful not write the stats and  returns false